storage/accumulator/src/lib.rs

```diff
@@ -181,34 +181,63 @@ where
         }
 
         let num_new_leaves = new_leaves.len();
-        let last_new_leaf_count = self.num_leaves + num_new_leaves as u64;
-        let root_level = Position::root_from_leaf_count(last_new_leaf_count).level() as usize;
+        let last_new_leaf_count = self.num_leaves + num_new_leaves;
+        let root_level = Position::root_level_from_leaf_count(last_new_leaf_count);
         let mut to_freeze = Vec::with_capacity(Self::max_to_freeze(num_new_leaves, root_level));
 
-        // create one new node for each new leaf hash
-        let mut current_level = self.gen_leaf_level(new_leaves);
-        Self::record_to_freeze(
-            &mut to_freeze,
-            &current_level,
-            false, /* has_non_frozen */
-        );
-
-        // loop starting from leaf level, upwards till root_level - 1,
-        // making new nodes of parent level and recording frozen ones.
-        let mut has_non_frozen = false;
-        for _ in 0..root_level {
-            let (parent_level, placeholder_used) = self.gen_parent_level(&current_level)?;
+        let mut num_leaves = self.num_leaves;
 
-            // If a placeholder node is used to generate the right most node of a certain level,
-            // such level and all its parent levels have a non-frozen right most node.
-            has_non_frozen |= placeholder_used;
-            Self::record_to_freeze(&mut to_freeze, &parent_level, has_non_frozen);
+        // Iterate over the new leaves, adding them to to_freeze and then adding any frozen parents
+        // when right children are encountered.  This has the effect of creating frozen nodes in
+        // perfect post-order, which can be used as a strictly increasing append only index for
+        // the underlying storage.
+        //
+        // We will track newly created left siblings while iterating so we can pair them with their
+        // right sibling, if and when it becomes frozen.  If the frozen left sibling is not created
+        // in this iteration, it must already exist in storage.
+        let mut left_siblings: Vec<(_, _)> = Vec::new();
+        for leaf in new_leaves.iter() {
+            let leaf_pos = Position::from_leaf_index(num_leaves as u64);
```

Comment:
You could do `for (leaf_offset, leaf) in new_leaves.iter().enumerate() {` above, do this here:
```suggestion
            let leaf_pos = Position::from_leaf_index(self.num_leaves + leaf_offset as u64);
```
and remove lines 188 and 203